pkg/rbac/decorator.go

```diff
@@ -0,0 +1,145 @@
+package rbac
+
+import (
+	"strings"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/orm"
+	"github.com/iov-one/weave/x"
+)
+
+const (
+	roleParticipantGasCost = 10
+)
+
+// AuthNDecorator handles authentication.
+type AuthNDecorator struct {
+	authN         x.Authenticator
+	roleBucket    orm.ModelBucket
+	roleBinBucket *RoleBindingBucket
+}
+
+func NewAuthNDecorator(auth x.Authenticator) AuthNDecorator {
+	return AuthNDecorator{
+		authN:         auth,
+		roleBucket:    NewRoleBucket(),
+		roleBinBucket: NewRoleBindingBucket(),
+	}
+}
+
+// Check enforces roles added to the context before calling down the stack
+func (d AuthNDecorator) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Checker) (*weave.CheckResult, error) {
+	newCtx, cost, err := d.authRoles(ctx, store)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := next.Check(newCtx, store, tx)
+	if err != nil {
+		return nil, err
+	}
+	res.GasPayment += cost
+	return res, nil
+}
+
+// Deliver enforces roles added to the context before calling down the stack
+func (d AuthNDecorator) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx, next weave.Deliverer) (*weave.DeliverResult, error) {
+	newCtx, _, err := d.authRoles(ctx, store)
+	if err != nil {
+		return nil, err
+	}
+
+	return next.Deliver(newCtx, store, tx)
+}
+
+func (d AuthNDecorator) authRoles(ctx weave.Context, db weave.KVStore) (weave.Context, int64, error) {
+	var costs int64
+	roleIdsProcessed := make(map[string]Role)
+	for _, c := range d.authN.GetConditions(ctx) {
+		roleIDs, err := d.roleBinBucket.FindRoleIDsByAddress(db, c.Address())
```

Comment:
okay, so you do start with an address and then load all roles it belongs to.

if a role was controlled by a multisig, then that must be in conditions before this decorator, but otherwise, I think this would work without changes.

Reply:
👍